### PR TITLE
Epetra-free build patch for issue #313

### PR DESCRIPTION
**Pull request by gahansen**
Changes were made to Stratimikos and Amesos2 to allow building
without Epetra - to support a lightweight Albany build.

These changes require a slight modification to Teko to cure a missing
header dependency created by the above changes.

The matching Trilinos cmake script is also included.

### CODE REVIEW

**Comment by eric-c-cyr** [comment]
I have no issues regarding Teko at this time.

File: packages/stratimikos/CMakeLists.txt [hunk 1]
```diff
@@ -6,6 +6,16 @@ TRIBITS_ADD_ENABLE_TEUCHOS_TIME_MONITOR_OPTION()
 
 TRIBITS_ADD_EXPLICIT_INSTANTIATION_OPTION()
 
+# Original behavior was to include ThyraEpetraAdapters as a required dependency.
+# Modify this to make sure that ThyraEpetraAdapters is enabled if Epetra is, and
+# that ThyraTpetraAdapters is enabled if Tpetra is
+IF(${PACKAGE_NAME}_ENABLE_Epetra )
+  SET(${PACKAGE_NAME}_ENABLE_ThyraEpetraAdapters ON)
```
Comment (bartlettroscoe) [inline packages/stratimikos/CMakeLists.txt]: Does this logic actually work?  I don't believe setting `${PROJECT_NAME}_ENABLE_<anything>=ON` has any effect (at at least not the effect you want) when you set this in a `<packageDir>/CMakeLists.txt`.  See [Full TriBITS Project Configuration](https://tribits.org/doc/TribitsDevelopersGuide.html#full-tribits-project-configuration).

**Comment by bartlettroscoe** [comment]
@gahansen, wow, is this really all you have to change to get Stratimikos to build, link, and and run correctly with Tpetra but without Epetra?

**Comment by bartlettroscoe** [comment]
@gahansen, from looking at your [configure script](https://github.com/gahansen/Trilinos/blob/f79eb403198c9b94ea3d959481a9a91593aef983/sampleScripts/gahanse-skybridge-epetrafree-release-cmake), it does not appear that any tests are enabled.  What happens when you turn the tests on for Stratimikos?  Since I think that all of that automated tests for Stratimikos rely on Epetra, I suspect that there will be no tests  if we update the files correctly.

Another issue I just thought about is that making this change will technically break backward compatibility.  That is, if there are customers that are simply enabling Straimikos with optional packages disabled, then they will not get any Epetra support for Stratimikos when they configure Trilinos.  Therefore, we may need to make this change conditional and then make it flip the switch when we drop backward compatibility for Trilinos 13.0.0-dev.  Therefore, we might need to protect this change in dependencies based on (perhaps) `<PackageName>_HIDE_DEPRECATED_CODE` (see [Setting up support for deprecated code handling](https://tribits.org/doc/TribitsDevelopersGuide.html#setting-up-support-for-deprecated-code-handling)) or create our own variable for this like `Stratimikos_MAKE_EPETRA_OPTIONAL` (which will have a default value of `FALSE`).  Then you would just need to configure Trilinos with  `-DStratimikos_MAKE_EPETRA_OPTIONAL=TRUE`.  Since this is not the standard build, we would need to set up a Nightly build of Trilinos that disables Epetra and sets `-DStratimikos_MAKE_EPETRA_OPTIONAL=TRUE` and made sure that all of the rest of these packages continued to work.

**Comment by ibaned** [comment]
I'm closing this in favor of #1023. Thanks @gahansen for figuring out what needs to be done, I'll carry on trying to get the remaining changes merged.

**Pull request closed without merging**